### DISCUSSION TITLE
Documentation

### DISCUSSION THREAD

**Discussion by amardeep** [gitext-rs/git-stack#100] [General]
TLDR: It would be great to improve documentation.

I have not used phabricator or other tools like that. In our team, we follow master based development, linear history (no merge commits). But we are often working on multiple features in parallel. As such, just using git-stack --push and git-stack --pull is a godsend. So thank you :) But without having any context of other such tools/processes, it is hard to follow the documentation. For eg:

- What is protected branch? What it should be (I could guess, but would be nicer to elaborate more)
- WIP: Would be nice to elaborate - what prefixes could be used - what happens when it is there.
- In Using section, there is a comment `# To push whats ready`. But what's considered ready
- Examples in the FAQ section - I am totally lost - can't understand them :(
- `git switch -v feature2` - I couldn't find -v flag for git switch (Maybe just a typo for -c)
- In faq "How do I add a commit to a branch (PR)?". Normally, I am working on a branch, and
   I just add commits on top of that branch. Here it says I have to use `git rebase -i`. Now I am
   wondering if by not using rebase I will break it. Or maybe you mean that if you want to move
   commit to a different branch - then there is no extra support - just do what you would have done
   and rebase.
- In "How do I start a new feature?", it says `git switch feature1 && git switch -c feature2`. Again,
   I am confused why is `git switch feature1` required. If I want to add feature on top of current changes,
   I would use `git switch -c feature2`. If I want to work on the feature as independent feature, I use
   `git switch -c feature2 origin/master --no-track`. If I were to adopt branchless workflow, perhaps I
   wouldn't worry about branches at all until I was ready to send for review.
- What does `git-stack --rebase` really do and in which scenarios it should be used.

**Comment by epage** [comment]
> What is protected branch? What it should be (I could guess, but would be nicer to elaborate more)

The definition got moved to "Examples".  This should be in a more clear place

> Parent branch auto-detection works by separating the concept of upstream-controlled branches (called "protected branches") and your development branches.

**Comment by epage** [comment]
Overall great feedback! Thank you for this!

**Comment by epage** [comment]
Mind looking over the docs again? I'm wondering how much my changes help and where they could use more attention

If you want to specifically see whats changed
- https://github.com/epage/git-stack/commit/4d12a7d6a1efa2767b5372bc60951e41b78e2ba5
- https://github.com/epage/git-stack/commit/7687724f1aef8c98b5b9cb3f738accaaa0fbdc42

**Reply by epage** [reply]
One thing I can see being added is an overview of the concepts, like I did in https://github.com/arxanas/git-branchless/discussions/125#discussioncomment-1505736.

Created https://github.com/epage/git-stack/issues/105

**Reply by amardeep** [reply]
These changes are very useful. Thanks.

Rewording faqs help a lot. First faq "How do I stack another branch on top of an existing one?" is still confusing for me - mainly as I don't understand what `git-stack --rebase` does. It would really help to have a separate section on that. Also in `git-stack --rebase --base HEAD~~ --onto feature` - is it supposed to be feature or one of feature[12]? Also, two bullet points are the same - "Moving existing". Do both need to be done or are they alternatives (I suspect once i understand what git-stack --rebase is doing it will make more sense)

What you have already is good enough for me, though i feel it would be nice if git-stack --pull and git-stack --push (also --rebase? ) had their own sections. Perhaps taking some content from post on git-branchless and docs/comparison.md. I think the content in comparison.md where you compare with raw git is very useful, and would be better if present in main README. It gives a better sense of what these commands are doing internally (as equivalent git concepts) and why it is worth using them.

Also, another section worth considering is if anything special should be done to uninstall it - does it keep some files in .git which could be deleted etc.

**Reply by epage** [reply]
> " is still confusing for me - mainly as I don't understand what git-stack --rebase does. It would really help to have a separate section on that

The second commit included that but I've expanded it further, including yoru idea of moving the raw git comparison to it

See https://github.com/epage/git-stack/blob/main/docs/reference.md#commands

> Also in git-stack --rebase --base HEAD~~ --onto feature - is it supposed to be feature or one of feature[12]? Also, two bullet points are the same - "Moving existing". Do both need to be done or are they alternatives (I suspect once i understand what git-stack --rebase is doing it will make more sense)

I've updated it to:

> - New branch: `git switch feature1 && git switch -c feature2` and start adding commits
> - Moving existing: `git stack --rebase --base feature1 --onto main` moves `feature2` to `main`, from off of `feature1`
>   - Without `git stack`: `git rebase feature1 --onto main`

> Also, another section worth considering is if anything special should be done to uninstall it - does it keep some files in .git which could be deleted etc.

Huh, hadn't thought of this because `git-stack` is pretty mundane in this way.  The only files we leave behind are the backups, and we cap how many backups we maintain.  I've added

> ### Uninstall
>
> See the uninstall method for your installer.
>
> Once removed, `git-stack` leaves behind:
> - `.git/branch-stash`
>
> Removing this is safe and will have no effect.

**Reply by amardeep** [reply]
Wow, all these changes look really good :) Especially the reference page. Thanks.

**Reply by epage** [reply]
Thank you! This feedback has been great!

**Comment by uloco** [comment]
Having also never worked with a branchless or stacked workflow myself, I am not quite sure how I should work with this tool. Especially if the rest of the team is still using a feature branch strategy with code reviews. It would really help if you could add more documentation in general, when to use it and what to achieve, what workflow you follow in your day to day work with it. Maybe some videos / screen casts would help a lot to get an overall understanding what this tool helps me in doing and what it doesn't.

I am constantly asking myself if I am doing this correctly, when to pull / push / rebase etc.
I have a long lived branch that is a migration, should I protect this also since a lot of branches are going into this? (for example)
If you have a good reference on this topic in general, feel free to share :).

I read the article in the Readme about the branchless workflow and it sounds great but I am at not point where I can convince my team to switch full on, so I need a more "we are using a standard branch workflow, what can I use this tool here for" point of view.

**Reply by epage** [reply]
> Especially if the rest of the team is still using a feature branch strategy with code reviews. 
> ...
> I read the article in the Readme about the branchless workflow and it sounds great but I am at not point where I can convince my team to switch full on, so I need a more "we are using a standard branch workflow, what can I use this tool here for" point of view.

There is a point of confusion here that I'm wondering what we can do to improve.  git-stack is specifically not for branchless development but focuses on the "stack" part of those articles.

Areas we currently try to make this clear:

Tag line:
> Stacked branch management for Git

About (emphasis on "bringing Stacked Diff workflow to PRs"):  
> Like Stacked-Diffs? git-stack is [another approach](https://github.com/gitext-rs/git-stack/blob/main/docs/comparison.md) to bringing the [Stacked Diff workflow](https://jg.gg/2018/09/29/stacked-diffs-versus-pull-requests/) to PRs that aims to be unintrusive to a project's workflow

About
> Unfamiliar with Stacked-Diffs? git-stack helps automate a lot of common workflows when dealing with PRs, especially when you start to create PRs on top of PRs.

The feature list includes branch related features

The [example workflow](https://github.com/gitext-rs/git-stack#example) talks in terms of branches and shows the equivalent with a branch workflow.

Any thoughts on how we could improve this to have prevented this confusion?

**Reply by epage** [reply]
> I am constantly asking myself if I am doing this correctly, when to pull / push / rebase etc.

For my personal workflow, I `git stack --pull` all the time.  I `git stack --push` when I'm ready to create or update a PR.  Before I do so, I sometimes run `git stack --fixup squash`.

I could expand the example with this.

**Reply by epage** [reply]
> I have a long lived branch that is a migration, should I protect this also since a lot of branches are going into this? (for example)

In the README (which is also expanded in the [reference](https://github.com/gitext-rs/git-stack/blob/main/docs/reference.md)
> Protected branches: These are branches like main or v3 that git-stack must not modify. 

If the branch is shared with others, it should be protected.  If its long-lived, it probably should be protected but that depends a lot on how you are using it.  Or put another way, would you be rebasing this branch yourself if you weren't using git-stack?

Any thoughts on how we can improve this to have avoided confusion?

**Reply by epage** [reply]
> I could expand the example with this.

I forgot the "Getting Started" as a more thorough workflow.  The first example is just to highlight what all is being done for you.

**Reply by uloco** [reply]
Thanks for your quick reply :)))

> Any thoughts on how we could improve this to have prevented this confusion?

I think the confusion comes here from the article talking a lot about branchless workflow while talking about stacked workflow. It mixes it up in the article so you (at least me) can get the idea, that you need to somehow do this to achieve the benefits of it.
I think it would be better to describe in your own words, what it means to have a stacked workflow ( for example branching on top of other branches)

**Reply by uloco** [reply]
> > I am constantly asking myself if I am doing this correctly, when to pull / push / rebase etc.
> 
> For my personal workflow, I `git stack --pull` all the time. I `git stack --push` when I'm ready to create or update a PR. Before I do so, I sometimes run `git stack --fixup squash`.
> 
> I could expand the example with this.

That's nice, that would be really great with some real world scenarios like this. :)

**Reply by uloco** [reply]
I was also confused about one thing. We usually create our merge requests directly from the gitlab ui, from an issue. So it creates a merge request and the according branch for it automatically. I then pull the branch to start working on it. Even when I base it off of a different branch.
I was confused if this would work as when I pull, I don't see any stacks (because there are no commits yet I guess) that was kind of confusing me if this workflow with not creating the branch locally first would work with git-stack.

**Reply by uloco** [reply]
> > I have a long lived branch that is a migration, should I protect this also since a lot of branches are going into this? (for example)
> 
> In the README (which is also expanded in the [reference](https://github.com/gitext-rs/git-stack/blob/main/docs/reference.md)
> 
> > Protected branches: These are branches like main or v3 that git-stack must not modify.
> 
> If the branch is shared with others, it should be protected. If its long-lived, it probably should be protected but that depends a lot on how you are using it. Or put another way, would you be rebasing this branch yourself if you weren't using git-stack?
> 
> Any thoughts on how we can improve this to have avoided confusion?

Just as you described it here it is way more clearer.

For example I was thinking that git-stack would understand, that main -> v3 -> feature structure, where main is the only protected branch. I was expecting git-stack to rebase v3 onto main and feature onto v3 when a new commit is added to main. but this did not happen because v3 was not protected at first. (maybe i also did something wrong, dunno...)

**Reply by uloco** [reply]
so i understood that i need to rebase v3 branch manually then... for now there are no other people working on the migration so rebases are kind of fine. afterwards I would just merge main into it when there are new changes and git-stack rebase my stacks.

**Reply by uloco** [reply]
Overall some real world scenarios like you did above would be great. For example, do you use stacks all the time or only for specific cases. I want to understand the overall workflow / thoughts on when to use it and how to use it on daily tasks.

**Reply by epage** [reply]
> I think it would be better to describe in your own words, what it means to have a stacked workflow ( for example branching on top of other branches)

The talk of stacked-diffs is to orient people coming from that background.  Rather than going into workflow theory, I give a summary for people not used to stacked-diffs:

> Unfamiliar with Stacked-Diffs? git-stack helps automate a lot of common workflows when dealing with PRs, especially when you start to create PRs on top of PRs.

**Reply by epage** [reply]
>  that would be really great with some real world scenarios like this. :)

As I said, further down we have one.  I am updating it a bit and cross-linking from the earlier example.

**Reply by epage** [reply]
> I was confused if this would work as when I pull, I don't see any stacks (because there are no commits yet I guess) that was kind of confusing me if this workflow with not creating the branch locally first would work with git-stack.

If you have your branch locally, it should be showing up on the same commit as your base branch.  This is assuming you have a local master/main.

**Reply by epage** [reply]
> Just as you described it here it is way more clearer.

I'm updating the description to

> **Protected branches:** These are branches that `git-stack` should not modify. `git-stack` will also rebase local protected branches against their remote counter parts.  Usually you mark shared or long-lived branches as protected, like `main`, `v3`.

> For example I was thinking that git-stack would understand, that main -> v3 -> feature structure, where main is the only protected branch. I was expecting git-stack to rebase v3 onto main and feature onto v3 when a new commit is added to main. but this did not happen because v3 was not protected at first. (maybe i also did something wrong, dunno...)

I'm a bit confused by this statement.  `v3` isn't marked as protected but `git-stack` didn't rebase it onto `main`?  That seems like a bug unless it reported a conflict.

**Reply by epage** [reply]
> Overall some real world scenarios like you did above would be great. For example, do you use stacks all the time or only for specific cases. I want to understand the overall workflow / thoughts on when to use it and how to use it on daily tasks.

Some people try to have one stack and just re-order items within that.  I worry about there being conflicts and instead organize my stacks by whether they are logically related and/or depend on each other.

I'll try to find some place to note this.

**Reply by epage** [reply]
See
- https://github.com/gitext-rs/git-stack/commit/6c2558f70e4285d69bf81e42c82682d1e7fb36e7
- https://github.com/gitext-rs/git-stack/commit/b3a25f79bb0c338853c6f29f83577b8c9ba5106f

**Reply by uloco** [reply]
> I'm a bit confused by this statement. `v3` isn't marked as protected but `git-stack` didn't rebase it onto `main`? That seems like a bug unless it reported a conflict.

I did run git stack --rebase and it rebased v3 onto main, but I was also expecting that it would rebase feature onto v3 afterwards. (Maybe I got an error there, but I don't remember sry...)
It is good to know what I could expect to happen when I run the commands. Maybe this is also a bit my confusion. I am not confident about what I am doing and what to expect 😅 

I am also not quite sure which of the both commands --repair and --rebase I should use in which scenario. It should rebase / repair on pull automatically right?

**Reply by uloco** [reply]
Thanks for all the clarification! :)

**Reply by epage** [reply]
> I am also not quite sure which of the both commands --repair and --rebase I should use in which scenario. It should rebase / repair on pull automatically right?

`--pull` implies `--rebase`.  `--repair` is meant to be used after a manual operation messes up your stacks in an attempt to try to fix your stacks.  I wonder if `--fix` might be better :).

**Reply by uloco** [reply]
So what's the difference between --rebase and --repair then? When do I need to run which?
Do i need to run --rebase or --repair when I manually commit to a parent branch for example?
Do I need to run --rebase or --repair when I pull manually without git stack --pull?

**Reply by epage** [reply]
`--rebase` does just that, rebases all of your branches onto the target branch.

`--repair` is meant to patch up manual rebase operations. For example, `git stack --rebase` doesn't stop to allow resolving conflicts but stops part way.  You then might have to use `git rebase` to resolve things.  This could lead to your stack being split into two.  `--repair` is a series of heuristics to help in situations like this to try to restore stacks.

**Reply by uloco** [reply]
I thought `git stack --rebase` would also rebase all my branches for commits in the main branch (and also descendants) but it does only for the current branch I am on. I guess this is what confused me most.

So when I have a main branch, add two commits, then branch from there, add two commits and again branch from there and add some commits. When someone changes something in main and i do git stack --pull it updates all branches. But when I commit directly to main and do git stack --rebase i have to check out every branch myself. I was expecting it to behave like pull...

**Reply by epage** [reply]
Say you have
- 123 upstream/main
- 234
- 345 main
- 456
- 567 A
- 678
- 789 B

Adding a commit to `main` is like adding a commit to `A`.  When we rebase, we do not know what you had intended and don't guess.  This is where `--repair` comes in.

With that said, generally our support for developing directly on `main` is pretty weak and subject to change and/or breakage (it might have stopped working in the last release)  While I try to keep it in mind, I'm not likely to make it a focus because I personally have found it to be problematic and find it best to always be working off of a branch.  If people want to share how it benefits their workflows so I can better understand other use cases, then I'm more open to putting more work into it.
